### PR TITLE
Backport of some of MSE related changes from wpe-2.22

### PR DESCRIPTION
**Pull request by emutavchi**
Original PR for wpe-2.22: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/697/

### CODE REVIEW

File: Source/WebCore/Modules/mediasource/MediaSource.cpp [hunk 1]
```diff
@@ -280,7 +280,8 @@ void MediaSource::completeSeek()
     // 4. Resume the seek algorithm at the "Await a stable state" step.
     m_private->seekCompleted();
 
-    monitorSourceBuffers();
```
Comment (eocanha) [inline Source/WebCore/Modules/mediasource/MediaSource.cpp]: The current monitorSourceBuffers() implementation already checks m_pendingSeekTime.isInvalid() before calling completeSeek() so, in theory, the recursion should already be stopped there and this patch shouldn't be needed.

Have you been able to reproduce the infinite recursion on 2.28? Do you have a backtrace to check?

File: Source/WebCore/Modules/mediasource/SourceBuffer.cpp [hunk 2]
```diff
@@ -798,7 +798,7 @@ static PlatformTimeRanges removeSamplesFromTrackBuffer(const DecodeOrderSampleMa
                 additionalErasedRanges.add(previousSample.presentationTime() + previousSample.duration(), erasedStart);
         }
 
-        auto endIterator = trackBuffer.samples.presentationOrder().findSampleStartingOnOrAfterPresentationTime(erasedEnd);
+        auto endIterator = trackBuffer.samples.presentationOrder().findSampleContainingOrAfterPresentationTime(erasedEnd);
```
Comment (eocanha) [inline Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: Can you explain a practical example where findSampleStartingOnOrAfterPresentationTime() causes problems and findSampleContainingOrAfterPresentationTime() solves them? Having a good practical case for justifying this change would be helpful in order to submit this change to WebKit upstream (our policy now for wpe-2.28 commits).

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h [hunk 3]
```diff
@@ -85,6 +85,8 @@ class MediaPlayerPrivateGStreamerMSE : public MediaPlayerPrivateGStreamer {
     WTFLogChannel& logChannel() const final { return WebCore::LogMediaSource; }
 #endif
 
+   bool gstSeekCompleted() const { return m_gstSeekCompleted; }
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:88]: Reminder to myself: This commit isn't upstreamable.

File: Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.h [hunk 4]
```diff
@@ -93,20 +93,24 @@ class MediaPlayerPrivateHolePunch : public MediaPlayerPrivateInterface, public C
 
     void pushNextHolePunchBuffer();
     void swapBuffersIfNeeded() final;
+    void setNetworkState(MediaPlayer::NetworkState);
 #if !USE(NICOSIA)
     RefPtr<TextureMapperPlatformLayerProxy> proxy() const final;
 #endif
 
+protected:
+    MediaPlayer* m_player;
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.h]: Why is this change to "protected" needed? Is there going to be a MediaPlayerPrivateHolePunch subclass somewhere?

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp [hunk 5]
```diff
@@ -2916,6 +2916,7 @@ void MediaPlayerPrivateGStreamer::createGSTPlayBin(const URL& url, const String&
 #endif
     }), this);
 
+    g_signal_connect_swapped(m_pipeline.get(), "element-setup", G_CALLBACK(elementSetupCallback), this);
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:2919]: Reminder to myself: I don't think this commit is upstreamable, as there's no support for Westeros sink upstream.

File: Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp [hunk 6]
```diff
@@ -134,10 +134,22 @@ static void enabledAppsrcNeedData(GstAppSrc* appsrc, guint, gpointer userData)
         // Search again for the Stream, just in case it was removed between the previous lock and this one.
         appsrcStream = getStreamByAppsrc(webKitMediaSrc, GST_ELEMENT(appsrc));
 
-        if (appsrcStream && appsrcStream->type != WebCore::Invalid)
-            webKitMediaSrc->priv->notifier->notify(WebKitMediaSrcMainThreadNotification::ReadyForMoreSamples, [webKitMediaSrc, appsrcStream] {
+        if (appsrcStream && appsrcStream->type != WebCore::Invalid) {
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:137]: This commit is useful, but is going to be superseeded by a patch @calvaris is working on, so let's better not merge it into wpe-2.28 yet.
See the original discussion about the topic in https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/791

**Review by eocanha — COMMENTED** [review]
Instead of directly merging this PR into wpe-2.28, I'm going to try to upstream the relevant commits and then backport them to wpe-2.28. I'll manually commit those commits that aren't upstreamable.
I've also left some questions for which it would be nice to get an answer. They would be helpful to make a case for upstreaming the related changes.

File: Source/WebCore/Modules/mediasource/MediaSource.cpp (hunk 1 above)
Comment (emutavchi) [reply Source/WebCore/Modules/mediasource/MediaSource.cpp]: Indeed. I missed that. Let me remove the commit from the PR.

**Comment by calvaris** [comment]
I might be interesting to file this kind of PRs in smallest chunks in the future.

**Comment by eocanha** [comment]
List of upstream commits corresponding to the ones in this PR:
- Bug [237526](https://bugs.webkit.org/show_bug.cgi?id=237526): [GStreamer] Handle gst pad error as format error
- Bug [237528](https://bugs.webkit.org/show_bug.cgi?id=237528): [MSE][SourceBuffer] Fix creating float PTS/DTS when dividing sample
- Bug [237780](https://bugs.webkit.org/show_bug.cgi?id=237780): [MSE][SourceBuffer] Fix removal of paddings from buffered range
- Not upstreamed: [MSE] Fix occasional video playback stalls after seek
- Bug [237602](https://bugs.webkit.org/show_bug.cgi?id=237602): [GStreamer] clarify playback errors with MediaError.message
- Bug [237828](https://bugs.webkit.org/show_bug.cgi?id=237828): [GStreamer] Avoid auto-selection of hole punching player
- Not upstreamed: [Westeros-sink] Initial playbin3 support
- Not upstreamed: Disable endcoded data replacement for images with async decoding
- Not upstreamed: Fix loss of ReadyForMoreSamples notifications causing video freeze

File: Source/WebCore/Modules/mediasource/SourceBuffer.cpp (hunk 2 above)
Comment (emutavchi) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: This change fixes high memory usage in apps that do TS -> MP4 transmuxing. The high memory usage is observed in some streams that have a slight overlap in media samples (usually happens on video quality change), when application applies its own logic for samples removal (that looks like sourceBuffer.remove(sourceBuffer.buffered.start(0), ...)). 

Current implementation reports a buffered range that is inconsistent with the actual samples that are still in the buffer (after range removal). The reported range excludes one sample, allowing that sample to "leak". Such a sample holds a reference to the entire appended chunk (due to in-place demuxing in GStreamer MSE backend), but it accounts only for a small size in the buffer. So eviction algorithm doesn't kick in.

Here you can find a mock test that simulates the problem: https://github.com/emutavchi/tests/blob/master/mse/media-source-remove.html. You'll need WebKit with Mock MSE and enable MSE debug logs to see the problem. Here is the output observed without change (range removal part):

```
[WPEWebKit:MediaSource:-] SourceBuffer::remove(29ACFEDF) start = {"value":0}, end = {"value":3.96}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) start = {"value":0}, end = {"value":3.96}          
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":0,"numerator":0,"denominator":1000,"flags":1},"dts":{"value":0,"numerator":0,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"dts":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":2,"numerator":2000,"denominator":1000,"flags":1},"dts":{"value":2,"numerator":2000,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":3,"numerator":3000,"denominator":1000,"flags":1},"dts":{"value":3,"numerator":3000,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removed 256, start = {"value":0,"numerator":0,"denominator":1000,"flags":1}, end = {"value":4,"numerator":4000,"denominator":1000,"flags":1}                                                                                                

[WPEWebKit:MediaSource:-] SourceBuffer::remove(29ACFEDF) start = {"value":4.29}, end = {"value":5}
[WPEWebKit:Media:-] HTMLMediaElement::scheduleUpdatePlayState(29ACFEDF) lambda(), task fired
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) start = {"value":4.29}, end = {"value":5}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":4.29,"numerator":4290,"denominator":1000,"flags":1},"dts":{"value":4.29,"numerator":4290,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":4.62,"numerator":4620,"denominator":1000,"flags":1},"dts":{"value":4.62,"numerator":4620,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":4.95,"numerator":4950,"denominator":1000,"flags":1},"dts":{"value":4.95,"numerator":4950,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removed 192, start = {"value":4.29,"numerator":4290,"denominator":1000,"flags":1}, end = {"value":5.28,"numerator":5280,"denominator":1000,"flags":1} 
```

note that sample at PTS 3.96 "leaked".

here is the output after the fix:
```
[WPEWebKit:MediaSource:-] SourceBuffer::remove(1FCB8181) start = {"value":0}, end = {"value":3.96}                                                                                                         
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) start = {"value":0}, end = {"value":3.96}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":0,"numerator":0,"denominator":1000,"flags":1},"dts":{"value":0,"numerator":0,"denominator":1000,"flags"
:1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"dts":{"value":1,"numerator":1000,"denominator":1000,"
flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":2,"numerator":2000,"denominator":1000,"flags":1},"dts":{"value":2,"numerator":2000,"denominator":1000,"
flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":3,"numerator":3000,"denominator":1000,"flags":1},"dts":{"value":3,"numerator":3000,"denominator":1000,"
flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removed 256, start = {"value":0,"numerator":0,"denominator":1000,"flags":1}, end = {"value":4,"numerator":4000,"denominator":1000,"flag
s":1}

[WPEWebKit:MediaSource:-] SourceBuffer::remove(1FCB8181) start = {"value":4}, end = {"value":5}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) start = {"value":4}, end = {"value":5}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":3.96,"numerator":3960,"denominator":1000,"flags":1},"dts":{"value":3.96,"numerator":3960,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":4.29,"numerator":4290,"denominator":1000,"flags":1},"dts":{"value":4.29,"numerator":4290,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":4.62,"numerator":4620,"denominator":1000,"flags":1},"dts":{"value":4.62,"numerator":4620,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":4.95,"numerator":4950,"denominator":1000,"flags":1},"dts":{"value":4.95,"numerator":4950,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removed 256, start = {"value":3.96,"numerator":3960,"denominator":1000,"flags":1}, end = {"value":5.28,"numerator":5280,"denominator":1
000,"flags":1}
```

File: Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp (hunk 6 above)
Comment (emutavchi) [reply Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:137]: @eocanha @calvaris are the changes you're working on will land any time soon? if not, I'd request this change to be merged.

File: Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.h (hunk 4 above)
Comment (emutavchi) [reply Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.h]: This was a straightforward cherry-pick. I don't remember why it was needed in 2.22 branch, but it looks like I can remove it in 2.28. I'll update the commit,

File: Source/WebCore/Modules/mediasource/MediaSource.cpp (hunk 1 above)
Comment (emutavchi) [reply Source/WebCore/Modules/mediasource/MediaSource.cpp]: @eocanha I've removed commit from PR.

File: Source/WebCore/Modules/mediasource/SourceBuffer.cpp (hunk 2 above)
Comment (jernoble) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: This result (sample at PTS 3.96 "leaked") is correct:

<https://www.w3.org/TR/media-source/#sourcebuffer-coded-frame-removal>

> Remove all media data, from this [track buffer](https://www.w3.org/TR/media-source/#track-buffer), that contain starting timestamps greater than or equal to start and less than the remove end timestamp.

Calling `sourceBuffer.remove(sourceBuffer.buffered.start(0), 3.96)` should not remove a sample whose presentation timestamp is `3.96` (and a sync sample) because that sample is not "less than the remove end timestamp".

Comment (emutavchi) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: Hi @jernoble, thanks for looking into this. The problem is with the buffered range report. The fix doesn't change removal behavior, it only affects the buffered range report. Currently, as you can see, the buffered range starts from 4.29 (after the first removal). This is not correct, 3.96 sample is still there in the buffer.

Comment (jernoble) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: Ah I see! Let me look more into why your proposed change has the effect you're trying to achieve and respond in a bit.

Comment (jernoble) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: So there's another problem with this content, in that the sample at `pts=3, duration=1` overlaps with the sample at `pts=3.96, duration=.33`. And that's causing a problem with calculating the end time of the removed range, which is what in turn causes this behavior.

I don't think that changing the method from `startingOnOrAfter` to `containingOnOrAfter` is actually the correct fix for this issue. The method was written with the assumption that samples in a given track do not overlap. After all, appending an overlapping sample _should_ evict the sample it overlaps with. I'm guessing there's a "fudge factor" somewhere that's preventing this eviction from happening.

A more correct fix would be to address why the samples are overlapping in the first place. If a fudge factor is causing a sample that would otherwise be evicted to not be, that sample should have its duration truncated (in the case of a video sample) to avoid the overlap. However, this is a much larger change, impacting much more than the bug you're trying to fix.

If that fix is not possible, the algorithm itself should be re-authored to take overlapping samples into account. For example, the loop immediately above the proposed change could only create a "erased range" spanning from the PTS of one sample to the PTS of the next, then expanding the erased range to the next (un-removed) sample's PTS.

Alternatively, if it is the case that the only way an overlapping sample is allowed to exist in the TrackBuffer is due to a fudge factor, another more correct fix would be to leave `findSampleStartingOnOrAfterPresentationTime()`, but decrement `erasedEnd` by `fudgeFactor`, to ensure you find the remaining sample.

Comment (jernoble) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: Whichever change you decide to go with, please feel free to post it on bugs.webkit.com and CC me; I'd be happy to work with you in getting a fix into the tree.

File: Source/WebCore/platform/graphics/MediaPlayer.cpp [hunk 7]
```diff
@@ -1229,7 +1229,7 @@ void MediaPlayer::networkStateChanged()
     // let the next engine try.
     if (m_private->networkState() >= MediaPlayer::NetworkState::FormatError && m_private->readyState() < MediaPlayer::ReadyState::HaveMetadata) {
         client().mediaPlayerEngineFailedToLoad();
-        if (!m_activeEngineIdentifier && installedMediaEngines().size() > 1 && (m_contentType.isEmpty() || nextBestMediaEngine(m_currentMediaEngine))) {
+        if (!m_activeEngineIdentifier && installedMediaEngines().size() > 1 && !m_contentType.isEmpty() && nextBestMediaEngine(m_currentMediaEngine)) {
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/MediaPlayer.cpp]: The original behaviour of loading with next media engine 
 in reloadTimerFired() when contentType is empty has been there since https://bugs.webkit.org/show_bug.cgi?id=58462, more than 10 years ago.
 Do you remember why this change was needed?

File: Source/WebCore/Modules/mediasource/SourceBuffer.cpp (hunk 2 above)
Comment (emutavchi) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: > So there's another problem with this content, in that the sample at pts=3, duration=1 overlaps with the sample at pts=3.96, duration=.33. And that's causing a problem with calculating the end time of the removed range, which is what in turn causes this behavior.

Right, this is what we observe on some live streams in various apps that do TS -> MP4 transmuxing. 

> I don't think that changing the method from startingOnOrAfter to containingOnOrAfter is actually the correct fix for this issue.

The end of the erased range would point either into an empty space or overlapped sample. if it points into an overlapped sample, I think it makes sense to not extend the erased range that would include that sample duration 

> Alternatively, if it is the case that the only way an overlapping sample is allowed to exist in the TrackBuffer is due to a fudge factor

I don't think sample replacement logic should kick in in this case (if I interpreted the MSE spec correctly).

> Whichever change you decide to go with, please feel free to post it on bugs.webkit.com and CC me; I'd be happy to work with you in getting a fix into the tree.

I still think that solution with containingOnOrAfter is good enough to fix the increased memory usage problem that we observe, without having to do a major rework.

File: Source/WebCore/platform/graphics/MediaPlayer.cpp (hunk 7 above)
Comment (emutavchi) [reply Source/WebCore/platform/graphics/MediaPlayer.cpp]: This is needed to notify html media element about the error, and preserve the player with an actual error message. otherwise, it just silently consumes the error while trying to load next engine.

File: Source/WebCore/Modules/mediasource/SourceBuffer.cpp (hunk 2 above)
Comment (jernoble) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: > > Alternatively, if it is the case that the only way an overlapping sample is allowed to exist in the TrackBuffer is due to a fudge factor
> 
> I don't think sample replacement logic should kick in in this case (if I interpreted the MSE spec correctly).

Except it does. Samples are not allowed to overlap with previously appended samples, but for various reasons (such as bad muxing or variant switching with different timebases) most UAs allow some amount of overlap. If we were strictly adhering to the spec, the sample at PTS=3.96 would evict the sample at PTS=3. 

The code to reconcile removed samples with buffered ranges looks at the end PTS time of those samples in order to define the removed range; if that removed sample (PTS=3) had its duration truncated to 3.96, the reconciliation algorithm would succeed normally.

> > Whichever change you decide to go with, please feel free to post it on bugs.webkit.com and CC me; I'd be happy to work with you in getting a fix into the tree.
> 
> I still think that solution with containingOnOrAfter is good enough to fix the increased memory usage problem that we observe, without having to do a major rework.

The problem with that solution is it does not take into account the amount of overlap. The solution proposed here would allow for situations that absolutely cannot exist, such a removed sample whose timeline is entirely within a non-removed sample, breaking not only the assumption that samples do not overlap, but that samples monotonically increase in both start PTS and end PTS. I would rather not add the concept if a more limited solution (or, IMO, a more correct one) exists.

Comment (emutavchi) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: > Except it does.

I must be missing somthing. so, I'm trying to understand how this can happen. Please bear with me.
I see 2 places in the spec that are relevant and could trigger removal of buffered samples:

\# 1
<pre>13. If last decode timestamp for track buffer is unset and presentation timestamp falls within the presentation interval of a coded frame in track buffer, then run the following steps: ... </pre>

The last decoded timestamp is set so we can ignore this case.

\# 2
<pre>14. Remove existing coded frames in track buffer:
If highest end timestamp for track buffer is not set:
    Remove all coded frames from track buffer that have a presentation timestamp greater than or equal to presentation timestamp and less than frame end timestamp.
If highest end timestamp for track buffer is set and less than or equal to presentation timestamp:
   Remove all coded frames from track buffer that have a presentation timestamp greater than or equal to highest end timestamp and less than frame end timestamp.
</pre>

the highest end timestamp is set to 4.0 when an overlapped sample is received. it is higher than pts of "arriving" sample (3.96). so this also doesn't trigger replacement.

I've added additionall logging to trace the processing of coded frames (https://github.com/emutavchi/WebKitForWayland/commit/f11ba3f2fb3124cf9bee3f2c53ceccabc7854e1a).

Here you can find full output: https://gist.github.com/emutavchi/47b4e488156a2018fecb4e533eef2603. 

Processing of sample at 3.96 PTS doesn't trigger removal:
```
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) {"pts":{"value":3.96,"numerator":3960,"denominator":1000,"flags":1},"dts":{"value":3.96,"numerator":3960,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) Last decode time stamp is valid, carry on
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) HighestPresentation timestamp is valid, {"value":4,"numerator":4000,"denominator":1000,"flags":1}
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) HighestPresentation timestamp is valid, no overlap found!
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) erasedSamples is empty
```

so, unless there is something else in the spec that would trigger replacement (and not implementet yet), I conclude that overlapping of media samples is allowed.

Comment (jernoble) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: > > Except it does.
> 
> I must be missing somthing. so, I'm trying to understand how this can happen. Please bear with me. I see 2 places in the spec that are relevant and could trigger removal of buffered samples:
> 
> # 1
> 
> 13. If last decode timestamp for track buffer is unset and presentation timestamp falls within the presentation interval of a coded frame in track buffer, then run the following steps: ... 
> The last decoded timestamp is set so we can ignore this case.

If a container is generating overlapping samples within a coded frame group, it can only be the result of either a bad mux or a demuxing error. The spec has an explicit requirement for monotonically increasing decode timestamps, and an implicit requirement for the sample end times to be well ordered. I'd have to check the ISO BMFF spec, but it should not be the case that samples within a container can be overlapping.

Your test case doesn't reset the (mock) parser between sample `PTS=3` and `PTS=3.96`. Does the in-the-wild behavior you're trying to fix reset the parser (with a new initialization segment) between appends?

Comment (jernoble) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: (As for the Mock test case output and logging, you could do all kind of weird things with the Mock "demuxer" and get strange behavior; including having non-monotonically increasing decode times, completely overlapping samples, etc. The Mock is just a tool to allow for cross-platform testing, that's all.)

Comment (emutavchi) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: I agree that the mock test is oversimplified. I'm sorry if that led to confusion. App does append init segment on quality change right before appending the overlapping samples.  

Here is an example that splits append into two and I think it'd better show the problem: https://raw.githubusercontent.com/emutavchi/tests/master/mse/media-source-remove2.html. I've also included a parser reset (abort) call before appending the second init segment (I'm not sure if the app did that, and I don't have logs from the reproduction case at the moment to confirm).

The behavior didn't change though, there was no sample replaced: https://gist.github.com/emutavchi/4820b0abd59f9e7645663b018d4b826f

Note that PTS and DTS are monotonically increasing.

Comment (emutavchi) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: And yes, I agree that this could be an issue with transmuxer on the web app side. but, we cannot fix that.

Comment (jernoble) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: I'd like to recognize that you are indeed correct; the spec does allow for overlapping video samples. I had misremembered; the behavior for audio and text samples is different than for video samples. Personally, I think that's a specification bug, and allowing samples to overlap in the TrackBuffer is confusing and will lead to errors like this one. I would support an addition to MSE that added a "video frame splicing" step similar to the audio and text splice frame algorithm which would simply truncate the duration of the previous sample.  But as the spec stands, that is not required.

That said, I don't think this affects what approach would be the most correct solution to the problem you are trying to solve. Given that overlapping frames can exist in the TrackBuffer, the algorithm's use of `erasedEnd` will inherently cause problems. I still think a more correct solution would be to re-write that section of the algorithm to search the TrackBuffer  using `findSampleStartingAfterPresentationTime()` and `erasedStart`.  Given that samples can overlap, using `findSampleContainingOrAfterPresentationTime()` with `erasedEnd` can still give incorrect results if multiple samples have presentation ranges containing that time.  Because that method uses a binary search algorithm, which sample is found in that scenario is non-deterministic. 

Alternatively, rather than using the sample's duration to calculate the `endTime` of the erased range, the algorithm could use the next remaining sample in PTS order to create the erased range. This would tackle the problem at the correct place, rather than adjusting the erased range to account for overlapping later in the algorithm.

Comment (emutavchi) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: I've updated as suggested to use findSampleStartingAfterPresentationTime() and erasedStart. https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/797/commits/880104d32c33cf39f12e42920a4f2bf25df2fd65

File: Source/WebCore/platform/graphics/MediaPlayer.cpp (hunk 7 above)
Comment (eocanha) [reply Source/WebCore/platform/graphics/MediaPlayer.cpp]: This caused regressions in Apple ports upstream, so I reverted the line to avoid player private regeneration and added code to cache the player private error in MediaPlayer as m_lastErrorMessage. That way, the error code of the disposed player private can be accessed after regeneration and the regressions are avoided.

File: Source/WebCore/Modules/mediasource/SourceBuffer.cpp (hunk 2 above)
Comment (eocanha) [reply Source/WebCore/Modules/mediasource/SourceBuffer.cpp]: I've ported the patch and the layout test upstream as https://bugs.webkit.org/show_bug.cgi?id=237780, just in case @jernoble wants to review it there.

File: Source/WebCore/loader/cache/CachedImage.cpp [hunk 8]
```diff
@@ -711,4 +711,11 @@ CachedResource::RevalidationDecision CachedImage::makeRevalidationDecision(Cache
     return CachedResource::makeRevalidationDecision(cachePolicy);
 }
 
+bool CachedImage::mayTryReplaceEncodedData() const {
```
Comment (eocanha) [inline Source/WebCore/loader/cache/CachedImage.cpp:714]: This commit is related to Graphics, away from my area of expertise (multimedia). Maybe @magomez might consider to upstream it. Otherwise I'm keeping it downstream only.

Comment (magomez) [reply Source/WebCore/loader/cache/CachedImage.cpp:714]: Async decoding was a relatively new feature in 2.22, but evolved a lot since then upstream, and this is probably not a problem anymore there. This should be kept as a downstream fix.

File: Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp (hunk 6 above)
Comment (calvaris) [reply Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:137]: I don't mind merging this now. I'll fix what I need to fix later.
